tests/run-make/reproducible-build/rmake.rs

```diff
@@ -0,0 +1,226 @@
+    // or remap path prefixes.
+    paths_test(PathsFlag::Link);
+    paths_test(PathsFlag::Remap);
+
+    // Builds should be reproducible even if each build is done in a different directory,
+    // with both --remap-path-prefix and -Z remap-cwd-prefix.
+
+    // FIXME(Oneirical): Building with crate type set to `bin` AND having -Cdebuginfo=2
+    // (or `-g`, the shorthand form) enabled will cause reproductibility failures.
+    // See https://github.com/rust-lang/rust/issues/89911
+
+    // This specific case would fail on OSX, should -Cdebuginfo=2 be added.
+    diff_dir_test(CrateType::Bin, RemapType::Path);
+
+    diff_dir_test(CrateType::Rlib, RemapType::Path);
+
+    //FIXME(Oneirical): This specific case would fail on both Linux and OSX, should -Cdebuginfo=2
+    // be added.
+    // See https://github.com/rust-lang/rust/issues/89911
+    diff_dir_test(CrateType::Bin, RemapType::Cwd { is_empty: false });
+
+    diff_dir_test(CrateType::Rlib, RemapType::Cwd { is_empty: false });
+    diff_dir_test(CrateType::Rlib, RemapType::Cwd { is_empty: true });
+
+    // Builds should be reproducible when using the --extern flag.
+    run_in_tmpdir(|| {
+        rustc().input("reproducible-build-aux.rs").run();
+        rustc()
+            .input("reproducible-build.rs")
+            .crate_type("rlib")
+            .extern_("reproducible_build_aux", rust_lib_name("reproducible_build_aux"))
+            .run();
+        rfs::copy(rust_lib_name("reproducible_build"), rust_lib_name("foo"));
+        rfs::copy(rust_lib_name("reproducible_build_aux"), rust_lib_name("bar"));
+        rustc()
+            .input("reproducible-build.rs")
+            .crate_type("rlib")
+            .extern_("reproducible_build_aux", rust_lib_name("bar"))
+            .run();
+        assert_eq!(rfs::read(rust_lib_name("foo")), rfs::read(rust_lib_name("reproducible_build")));
+    });
+}
+
+#[track_caller]
+fn smoke_test(flag: Option<SmokeFlag>) {
+    run_in_tmpdir(|| {
+        rustc().input("linker.rs").opt().run();
+        rustc().input("reproducible-build-aux.rs").run();
+        let mut compiler1 = rustc();
+        let mut compiler2 = rustc();
+        if let Some(flag) = flag {
+            match flag {
+                SmokeFlag::Debug => {
+                    compiler1.arg("-g");
+                    compiler2.arg("-g");
+                }
+                SmokeFlag::Opt => {
+                    compiler1.opt();
+                    compiler2.opt();
+                }
+            };
+        };
+        compiler1
+            .input("reproducible-build.rs")
+            .linker(&cwd().join(bin_name("linker")).display().to_string())
+            .run();
+        compiler2
+            .input("reproducible-build.rs")
+            .linker(&cwd().join(bin_name("linker")).display().to_string())
+            .run();
+        diff().actual_file("linker-arguments1").expected_file("linker-arguments2").run();
```

Comment:
Suggestion: at assertion sites, such as `diff().run()`, or `assert`s, can you please add an `eprintln!` which specifies which combination is hitting the assertion to aid debugging? Something like

```rs
eprintln!("smoke_test => SmokeFlag::Debug");
```

and

```rs
eprintln!("diff_dir_test => CrateType::Rlib, RemapType::Cwd { is_empty: true }");
```